### PR TITLE
container works on quest

### PR DESCRIPTION
**Pull request by danrlu**
1. to get the docker container work on Quest, need (1) do `module add singularity/latest` in terminal before starting nextflow run. For some reason this line just doesn't work properly inside of config (2) inside of config: `container = 'docker://andersenlab/gatk4:latest'`

2. I added `-XX:ConcGCThreads=${task.cpus}"` back to `gatk HaplotypeCaller` as fail safe since the JAVA garbage collection is multi threaded and may exceeds available resource and throw error.

3. The resource for sm, md, lg in quest.config was adjusted based on available resources on Quest and the actual resource usage from timeline.html generated by nextflow. Please do not modify them.

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -233,7 +234,7 @@ process call_variants_individual {
         tuple strain, path("${region}.g.vcf.gz")
 
     """
-        gatk HaplotypeCaller --java-options "-Xmx${task.memory.toGiga()}g -Xms1g" \\
+        gatk HaplotypeCaller --java-options "-Xmx${task.memory.toGiga()}g -Xms1g -XX:ConcGCThreads=${task.cpus}" \\
```
Comment (danielecook) [inline main.nf:237]: Why are you allocating additional threads for garbage collection? Does this improve performance?

Comment (danrlu) [reply main.nf:237]: i think it's from here: https://gatk.broadinstitute.org/hc/en-us/articles/360035532372-Java-is-using-too-many-resources-threads-memory-or-CPU-
gatk4 does not have built-in multi threading, they moved all that functionality to depend on spark. so it actually does not help too much to allocate gatk multiple cores. but, java has threads of its own, so this is telling java not to use more than there is available. i forgot whether this to deal with an error or warning.

**Comment by danielecook** [comment]
> to get the docker container work on Quest, need (1) do module add singularity/latest in terminal before starting nextflow run. For some reason this line just doesn't work properly inside of config (2) inside of config: container = 'docker://andersenlab/gatk4:latest'

The container should be specified in `nextflow.config` as it is used locally (for testing) and on quest. I'm not sure why you need the `docker://` prefix (it's probably better to leave it), but it may because you are using a different nextflow version as you commented out the `NXF_VER` check. I would recommend you leave that in as it ensures that the version used is consistent. You can use [direnv](https://direnv.net/) on quest to set variables at the directory level automatically. when you enter a directory.

> I added -XX:ConcGCThreads=${task.cpus}" back to gatk HaplotypeCaller as fail safe since the JAVA garbage collection is multi threaded and may exceeds available resource and throw error.

ok I would not think this would make a difference. Do you see jobs failing less often with it enabled? The JVM already sets this automatically. Generally, I would think you set it to restrain the number of GC threads.

> The resource for sm, md, lg in quest.config was adjusted based on available resources on Quest and the actual resource usage from timeline.html generated by nextflow. Please do not modify them.

None of those changes were committed. Values were only modified for testing purposes.

**Comment by danrlu** [comment]
> I'm not sure why you need the docker:// prefix (it's probably better to leave it), but it may because you are using a different nextflow version as you commented out the NXF_VER check. 

`docker://` tells nextflow to pull from docker hub. it is essential.
https://www.nextflow.io/docs/latest/singularity.html
image

I use nextflow 20.01.0. It is controlled by conda, instead of having `NXF_VER` setup so I commented that line out. 

I added back the `-Xmx${task.memory.toGiga()-3}g -Xms${task.memory.toGiga()-4}g` for `import_genomics_db`. reason see my last pull request.

**Comment by danielecook** [comment]
With regard to Nexflow, I have it installed with [homebrew](https://brew.sh).  This allows you to set the version more flexibly without needing to manage installation with condo. I would suggest you enforce some versioning if multiple people will be running the same pipeline.

**Comment by danrlu** [comment]
Got it! Thanks!

**Pull request merged**